.gitlab-ci.yml

```diff
@@ -0,0 +1,188 @@
+stages:
+  - protos
+  - build
+  - release
+  - test
+
+variables:
+  GIT_DEPTH: 1
+  RELEASE_BRANCH: develop
+  GRPC_TAG: $CI_REGISTRY/couchers/grpc:latest
+  PROXY_IMAGE: $CI_REGISTRY_IMAGE/proxy:latest
+  BACKEND_IMAGE: $CI_REGISTRY_IMAGE/backend:latest
+  MEDIA_IMAGE: $CI_REGISTRY_IMAGE/media:latest
+  FRONTEND_IMAGE: $CI_REGISTRY_IMAGE/frontend:latest
+  PROXY_TAG: $CI_REGISTRY_IMAGE/proxy:$CI_COMMIT_REF_SLUG-$CI_COMMIT_SHORT_SHA
+  BACKEND_TAG: $CI_REGISTRY_IMAGE/backend:$CI_COMMIT_REF_SLUG-$CI_COMMIT_SHORT_SHA
+  MEDIA_TAG: $CI_REGISTRY_IMAGE/media:$CI_COMMIT_REF_SLUG-$CI_COMMIT_SHORT_SHA
+  FRONTEND_TAG: $CI_REGISTRY_IMAGE/frontend:$CI_COMMIT_REF_SLUG-$CI_COMMIT_SHORT_SHA
+
+default:
+  image: docker
+  services:
+    - docker:19.03.12-dind
+  before_script:
+    - docker login -u $CI_REGISTRY_USER -p $CI_REGISTRY_PASSWORD $CI_REGISTRY
+
+protos:
+  stage: protos
+  image: $GRPC_TAG
+  inherit:
+    # the grpc container doesn't have docker, no need to login
+    default: false
+  script:
+    - cd app && ./generate_protos.sh
+  rules:
+    - changes:
+      - app/
+  artifacts:
+    paths:
+      - app/
+
+build:proxy:
+  stage: build
+  script:
+    - docker build -t $PROXY_TAG app/proxy/
+    - docker push $PROXY_TAG
+  rules:
+    - if: '$CI_COMMIT_BRANCH == $RELEASE_BRANCH'
+      changes:
+      - app/
+    - if: '$CI_COMMIT_BRANCH != $RELEASE_BRANCH'
+      changes:
+      - app/pb/
+      - app/proxy/
```

Comment:
docker automatically fast-forwards if none of the ADD:ed files are changed, so this section can be removed.

Reply:
I added these because the caching is not very solid on the CI system (it runs on different machines and it's not always shared, etc), I'm not even sure if it works at all currently (and you know... it's docker).

Reply:
yep, thanks for the explanation!